fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/TypeCoercion.java

```diff
@@ -0,0 +1,394 @@
+        }
+        if (expected instanceof TypeCollection) {
+            TypeCollection typeCollection = (TypeCollection) expected;
+            // use origin datatype first. use implicit cast instead if origin type cannot be accepted.
+            return Stream.<Supplier<Optional<DataType>>>of(
+                    () -> typeCollection.getTypes().stream()
+                            .filter(e -> e.acceptsType(input))
+                            .map(e -> input)
+                            .findFirst(),
+                    () -> typeCollection.getTypes().stream()
+                            .map(e -> implicitCast(input, e))
+                            .filter(Optional::isPresent)
+                            .map(Optional::get)
+                            .findFirst())
+                    .map(Supplier::get)
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .findFirst();
+        }
+        if (input.isNullType()) {
+            // Cast null type (usually from null literals) into target types
+            returnType = expected.defaultConcreteType();
+        } else if (input.isNumericType() && expected instanceof DecimalType) {
+            // If input is a numeric type but not decimal, and we expect a decimal type,
+            // cast the input to decimal.
+            returnType = DecimalType.forType(input);
+        } else if (input.isNumericType() && expected instanceof NumericType) {
+            // For any other numeric types, implicitly cast to each other, e.g. bigint -> int, int -> bigint
+            returnType = (DataType) expected;
+        } else if (input.isStringType()) {
+            if (expected instanceof DecimalType) {
+                returnType = DecimalType.SYSTEM_DEFAULT;
+            } else if (expected instanceof NumericType) {
+                returnType = expected.defaultConcreteType();
+            }
+        } else if (input.isPrimitive() && !input.isStringType() && expected instanceof CharacterType) {
+            returnType = StringType.INSTANCE;
+        }
+
+        // could not do implicit cast, just return null. Throw exception in check analysis.
+        return Optional.ofNullable(returnType);
+    }
+
+    private boolean canHandleTypeCoercion(DataType leftType, DataType rightType) {
+        if (leftType instanceof DecimalType && rightType instanceof NullType) {
+            return true;
+        }
+        if (leftType instanceof NullType && rightType instanceof DecimalType) {
+            return true;
+        }
+        if (!(leftType instanceof DecimalType) && !(rightType instanceof DecimalType) && !leftType.equals(rightType)) {
+            return true;
+        }
+        return false;
+    }
+
+    // TODO: compatible with origin planner and BE
+    // TODO: when add new type, add it to here
+    @Developing
+    private Optional<DataType> findTightestCommonType(DataType left, DataType right) {
```

Comment:
Missing Date/Datetime?

Reply:
yes, will add it in next PR about type coercion